src/inmanta/agent/agent.py

```diff
@@ -211,6 +211,25 @@ async def execute(
             # Explicit cast is required because mypy has issues with * and generics
             results: List[ResourceActionResult] = cast(List[ResourceActionResult], await asyncio.gather(*waiters))
 
+            if self.undeployable:
+                self.running = True
+                try:
+                    if self.is_done():
+                        # Action is cancelled
+                        self.logger.log(const.LogLevel.TRACE.to_int, "%s %s is no longer active" % (self.gid, self.resource))
+                        return
+                    result = sum(results, ResourceActionResult(cancel=False))
+                    if result.cancel:
+                        # self.running will be set to false when self.cancel is called
+                        # Only happens when global cancel has not cancelled us but our predecessors have already been cancelled
+                        return
+                    self.status = self.undeployable
+                    self.change = const.Change.nochange
+                    self.changes = {}
+                    self.future.set_result(ResourceActionResult(cancel=False))
```

Comment:
I don't really understand why you put this in a separate block, rather than to keep the condition you had before. Then we'd only need these four lines and the return part, rather than having to duplicate the whole block above.

Reply:
We need to make sure that we don't set the result of the future twice, as this would result in an exception. Setting `running` to True and checking the value of `is_done()` and `cancel` are required to make sure this doesn't happen. Strictly speaking it's not required to set `running` to True here because the code is entirely synchronous, but I added it here anyway for safety reasons. If we add an asynchronous call to this try-finally block in the future, we will probably not remember to add the try-finally block and set `running` to True...

Reply:
I understand that part, it's just that between this branch of the condition and the other there seems to be only a small difference. The whole complexity surrounding it with the `is_done` and `cancel` is duplicated here and I see no reason for that. Am I missing something?

Reply:
If it's not clear what I mean I can provide a rough diff.

Reply:
I think I understand your question. I mainly did that to bypass the rate limiter of the agent, because there is noreason why we would wait for it.

Reply:
Ok, that's a good point.